is_convergent method implementation for Sum

gxyd opened the pull request:
TODO
- [x] Divergence test
- [x] root test
- [x] comparison test
- [x] integral test
- [x] absolute convergence
- [ ] ratio test (should not be implemented in wake of [this](https://en.wikipedia.org/wiki/Convergence_tests#Comparison) ).
- [x] tests for `_order` added

Update
- [x] Alternating series test (finding some difficulties in implementing)

Commit:
is_convergent function implementation

is_convergent added to `Sum` class

doctest fixes

tests more

gxyd commented:
@jksuom

jksuom commented on sympy/calculus/singularities.py:
```diff
@@ -187,3 +187,93 @@ def is_monotonic(f, interval=S.Reals):
+    ==========
+
+    infinite_series : Expr
+            An expression whose convergence is to be checked
```
The examples below seem to indicate that ``infinite_series`` will not be a series but a term of a series. For the convergence of an infinite series it is necessary but not sufficient that the term sequence converge to zero. (That can be used as a first test to show that a series cannot be convergent.)

There is more to an infinite series than the general expression of a term and the index symbol, for example the limits that need not always be 1 and ``oo``. These could be added to the interface, but perhaps it would be better to have the whole series expression as the argument. Implemented that way ``is_convergent`` could become a method of the ``Sum`` class. This would make it easier to implement ``is_convergent`` for other classes as well.

gxyd commented:
I have added it to the `class Sum` method. But still the point i mentioned on gitter, regarding
```
>>> n = Symbol('n', integer=True)
>>> Sum(1/(n**3 - 1), (n, 0, oo)).is_convergent()
False
```
To me this looks good but i am not sure for this behavior.

jksuom commented:
I think ``Sum`` should somehow indicate that it does not make sense since the term with ``n == 1`` is not defined. (Actually SymPy gives ``zoo``, but that is not good here. We should only have finite terms.)

I am not sure what should be returned, if the construction of ``Sum`` is allowed (as it seems to be). Perhaps ``False`` could be the best choice at least for now.

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,93 @@ def _eval_simplify(self, ratio, measure):
+        Convergence tests are used for checking the convergence of an
+        a series. There are various tests employed to check the
+        convergence. Like divergence test, root test, integral test,
+        alternating series test, comparison tests.
```
Note that it returns `True` for convergent, `False` for divergent, and `None` for unknown.

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,93 @@ def _eval_simplify(self, ratio, measure):
+        try:
+            for singul in _sing:
+                if singul.is_integer and singul in _interval:
+                    return S.false
```
I think you should use regular True and False. S.true should only be used in places where a general symbolic boolean could be returned. Also, since this uses three-valued logic (with None), it should use `bool`, since S.true only represents two-valued logic.

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -790,3 +881,24 @@ def eval_sum_hyper(f, i_a_b):
         return Piecewise(res, (old_sum, True))
+
+
+def _order_growth(expression):
```
Why can't you just use `O`? The algorithms in O are going to be much smarter than this.

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -790,3 +881,24 @@ def eval_sum_hyper(f, i_a_b):
+    if expression.is_number:
+        return S.Zero
+
+    if expression.is_rational_function() and expression.as_numer_denom()[1] == 1:
```
Can this even happen?

asmeurer replied:
I guess this is the same as checking if it is a polynomial.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,93 @@ def _eval_simplify(self, ratio, measure):
+
+        ###  ------- Divergence test ---------- ###
+        try:
+            if limit(_sequence_term, sym, oo) != S.Zero:
```
Perhaps we should also consider series with lower limit ``-oo``.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,93 @@ def _eval_simplify(self, ratio, measure):
+        ### ------------ integral test ------------- ###
+        if is_decreasing(_sequence_term, _interval):
+            intgral_val = integrate(_sequence_term, (sym, _lower_limit, _upper_limit))
+            if intgral_val.is_infinite:
```
What would happen if the integral could not be evaluated? (Perhaps ``intgral_val`` could be just ``integral``.)

jksuom commented:
I have tried to discover the reason for the ample use of ``_`` prepended to the names of local variables. I have understood that they are often used to indicate that an attribute of an object is intended for internal use only. This does not seem to be the case here.

Commit:
Order function used

gxyd commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -855,3 +855,39 @@ def test_indexed_idx_sum():
+    # currently raise TypeError
+    assert Sum(sin(n)/n**3, (n, 1, oo)).is_convergent() == S.true
+    assert Sum(ln(n)/n**3, (n, 1, oo)).is_convergent() == S.true
+    assert Sum(1/(n**2*log(n)), (n, 2, oo)).is_convergent() == S.true
```
This works now, since of initial `Order` not able to handle i had added this. I will add to normal tests.

gxyd commented:
@asmeurer i have addressed your points regarding growth of function (`order_growth`), no longer needed.

gxyd commented:
@jksuom I have addressed your points.
> I have tried to discover the reason for the ample use of _ prepended to the names of local variables. I have understood that they are often used to indicate that an attribute of an object is intended for internal use only. This does not seem to be the case here.

Before this i had thought this to be a good practise, to always use `_variable_name`, compared to `variable_name`. But as you mentioned, i have removed `_` from variables.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +258,105 @@ def _eval_simplify(self, ratio, measure):
+
+        ### ---------- root test --------------- ###
+        lim = limit(abs(sequence_term)**(1/sym), sym, oo)
+        if lim < 1:
```
This is not valid for alternating series. The limit could be less than -1.

gxyd replied:
I do not understand your argument. The [wikipedia article](https://en.wikipedia.org/wiki/Root_test) on root_test states.
The root test states that:
`C = limit(sup(abs(a_n)**(1/n)), n -> oo)`
if `C < 1` then the series converges absolutely,
if `C > 1` then the series diverges,
if `C = 1` and the limit approaches strictly from above then the series diverges,
otherwise the test is inconclusive (the series may diverge, converge absolutely or converge conditionally).

jksuom replied:
Yes, of course. I missed that ``abs``. The limit is never negative.

jksuom replied:
I was thinking ot the possibility of taking the alternative case first. The ``abs`` might not be needed.

gxyd replied:
> I was thinking ot the possibility of taking the alternating case first. Then abs might not be needed.

Yes, you are right i guess for that `abs` would not be needed. But i think `root test` is more of a general test, so i think it should be there before the `alternating series test`.(Though i have not strict order of placing tests in mind).

jksuom replied:
I think it is better this way. ``root_test`` is more general and it works with series not necessarily alternating but still have terms of both signs.

asmeurer replied:
Don't use `if` on a symbolic boolean. You should factor this so that even if `lim < 1` is unevaluated it gives a correct answer.

gxyd commented:
@jksuom Do we have any `standard order of growth` above which the functions are "divergent" and below which "convergent"? For example `O(1/(n*log(n)), (n, oo)) < O(1/n, (n, oo))`. Now initially i thought that any function with `O(expr, (n, oo)) < O(1/n, (n, oo))` would be convergent. But i realised that is wrong, since `O(1/(n*log(n)) < O(1/n, (n, oo))` but `O(1/(n*log(n), (n, oo))` is "not convergent".
Sorry. By `<` operator here, i refer to `contains` actually.

gxyd commented:
Ya, i got it. That's was a naive question.

jksuom commented:
These series are good examples of the use of the integral test. One can say that there is some kind of pattern even if it is not fully general. Namely,

-  ``1/n**e``  convergent for ``e > 1``, divergent for ``e == 1`` (and ``e < 1``).
-  ``1/n*(log(n))**e`` is convergent for ``e > 1``, divergent for ``e == 1``.
-  ``1/n*log(n)*(log(log(n)))**e is convergent for ``e > 1``, divergent for ``e ==1``.
-  etc..

(Consider the derivatives of ``x**(1-e)``, ``log(x)``, ``log(x)**(1-e)``, ``log(log(x))``, etc...)

asmeurer commented:
This could be made to work better with symbolic parameters quite easily in most cases, I think, by using Piecewise instead of `if` statements.

asmeurer commented:
What about the (almost obvious) test of calling `doit()` on the Sum and checking if the result is finite?

asmeurer commented:
Although on second thought, how well does Piecewise work with boolean expressions? Maybe it should be returning a Boolean object (meaning go back to `S.true` and so on), but for something like `Sum(x**n, (n, 1, oo))` return `abs(x) < 1` (instead of `Piecewise((True, abs(x) < 1), (False, True))`).

asmeurer commented:
Of course, what you have here is a good start, so if you want to hold off on worrying about symbolic parameters, that's fine. Just make sure that they don't break the function (they should raise NotImplementedError or give None instead of raising TypeError).

jksuom commented:
I doubt if ``doit()`` on the ``Sum`` would be useful. On the integral it might work.

gxyd commented:
> I doubt if doit() on the Sum would be useful. On the integral it might work.

I agree with this. Most of the functions for which `Sum` is evaluated the `integral` is definitely evaluated for them, but the vice-versa may not be true. So that leaves with not much of its use.

Commit:
comparison test for log's added

gxyd commented:
@jksuom I have removed the `singularities` involved in `summations`, i have mentioned it on  my gitter comment.

Commit:
absolute_convergence for Sum

tests for log's added

matching 1/(n*log(n)**p) added

S.Infinity instead of oo

gxyd commented:
I have now added the `absolute convergence` for the `Sum`. @jksuom Please have a look.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +258,157 @@ def _eval_simplify(self, ratio, measure):
+            return False
+
+        # (1/(n*log(n)**p)) comparison
+        log_n_test = sequence_term.match((1/(sym*(log(sym))**p)))
```
I have no experience with matching, but I suspect that this should be more general to take care of terms like ``2/(n*log(n)**2)``. (Numerator different from one.)

Besides, this only works in a very special case and is more suitable for integral test. A comparison would be needed e.g. if the denominator would have an additional term of lower order (or growth), but that is harder to implement. There could be (integral) tests for these "log" cases.

gxyd commented:
> I have no experience with matching, but I suspect that this should be more general to take care of terms like 2/(n*log(n)**2). (Numerator different from one.)

Though one thing to mention that example `2/(n*log(n)**2)` was working fine as well at in the previous commit. Now, I have modified the `1/(n*log(n)*log(log(n))**p)` comparison-tests  such that it compares with `order of sequence_term` not with `sequence_term` , since `2/(n*log(n)*log(log(n))**p)` has `order` of `1/(n*log(n)*log(log(n))**p)` making the comparison a valid one.
I have removed the comparison for `1/(n*log(n)**p)`, since as you mentioned that is better handled by `Integral tests`.
I have added tests for `Integral tests` i.e the ones for `log's` you mentioned.

jksuom commented:
The situation has changed now that you match the order instead of the term. (I should have studied the workings of ``O`` more carefully. It seems that it is more capable than what I had expected.) With that modification even the old test (match ``1/(n*log(n)**p)``) would be preferable to the integral test. It can take care of cases where integration might not succeed. (Though it does with plain `1/(n*log(n)**e)``, of course.)

gxyd commented:
> I should have studied the workings of O more carefully.

Can you also refer me to sources for that., probably online :)
Apart from [this](https://gitter.im/sympy/sympy?at=55f3efd15ba1e0ea6b80cdb2) comment anything more to be addressed?

jksuom commented:
It may not be quite simple to improve ``O``. It is principally based on ``as_leading_term`` that assumes the existence of a series expansion. This true usually for expansions  at finite points but often not at infinities. It seems that (nearly?) all support currently consists of returning the function unevaluated. That gives sometimes correct results (as for ``O(1/(x*log(x)**t), (x, oo))``) but also useless answers as in
```
In [47]: print(O(1/(x*sin(x) + 1), (x, oo)))
O(1/(x*sin(x)), (x, oo))
```
The point is that ``O`` should only return functions that do not have negative values in a neighbourhood of the limit point. So ``sin(x)`` is wrong. My idea was to replace that with a positive valued majorant function (in this case the constant 1) at some place in the evaluation of order. That would help in cases like ``O(x*sin(x), (x, oo))``, but in the example there is no reasonable answer as the denominator has infinitely many zeroes.

Anyway, these are just some preliminary observations. I really haven't taken much time to study this area.

gxyd commented:
Here is the issue #5832 .

Commit:
Dirichlet test added

condition for Dirichlet test added

gxyd commented:
@jksuom Have a look. I have added the Dirichlet tests, which now handle the `sin(n)/n` convergence and made some changes to `order.expr`. Let me know are there any other `convergence-tests` or expressions which are not handled.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,174 @@ def _eval_simplify(self, ratio, measure):
+        dict_val = sequence_term.match((-1)**p*q)
+        if dict_val[p] == sym or dict_val[p].has(sym):
+            if solveset((dict_val[q].subs({sym: d + 1})/dict_val[q].subs({sym: d})) < 1, d, interval) == interval and \
+                limit(dict_val[q], sym, S.Infinity).is_finite:
```
I think this is superfluous. Actually, the limit should be 0, and it was tested already.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,174 @@ def _eval_simplify(self, ratio, measure):
+
+        ### --------- p-series test (1/n**p) ---------- ###
+        p1_series_test = order.expr.match(sym**p)
+        if not p1_series_test is None:
```
I think the recommended form is ``if ... is not ...``. (No difference of meaning but more readable.)

gxyd replied:
Yes, that also looks good.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,174 @@ def _eval_simplify(self, ratio, measure):
+        d = symbols('d', cls=Dummy)
+        dict_val = sequence_term.match((-1)**p*q)
+        if dict_val[p] == sym or dict_val[p].has(sym):
+            if solveset((dict_val[q].subs({sym: d + 1})/dict_val[q].subs({sym: d})) < 1, d, interval) == interval and \
```
Is is possible to use ``is_decreasing`` here?

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,174 @@ def _eval_simplify(self, ratio, measure):
+
+        ### -------------- Dirichlet tests -------------- ###
+        a_n = order.expr.args[0]
+        b_n = order.expr.args[1]
```
Should we have to verify some assumptions about the type of ``order.expr``?

gxyd replied:
I have the added the `.is_Mul` assumption checker. I am not sure if any more would be required.

jksuom commented:
I think this is (almost) ready to go. Only the Dirichlet test may be troublesome, but it is not essential. There are other, even more refined tests, but I don't think it would be necessary to include them now.

gxyd commented:
We have got
```
>>> Sum(sin(n**2)/n, (n, 1, oo)).is_convergent()   # this is handled by Integral-test
NotImplementedError
```
The thing is `integrate(sin(n**2)/n, (n, 1, oo)).is_finite` returns `None`, even as it is a `finite_number`, the thing goes to `Si(1).is_finite` returns `None`.

jksuom commented:
I don't think ``Sum(sin(n**2)/n, (n, 1, oo))`` should be handled by the integral test since ``sin(n**2)/n`` is not decreasing.

Commit:
Piecewise functions convergence added

test case for Piecewise added and un-necessary test removed

changes to piecewise function

Infinity to NegativeInfinity convergence handling corrected

Docs corrected

dirichlet for swapped arguments

gxyd commented:
> I don't think Sum(sin(n**2)/n, (n, 1, oo)) should be handled by the integral test since sin(n**2)/n is not decreasing.

Yes, you are right. It is indeed handled by `dirichlet test`.(i forgot to see the `Error` in SymPy)

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,181 @@ def _eval_simplify(self, ratio, measure):
+        # Piecewise function handle
+        if sequence_term.is_Piecewise:
+            for func_cond in sequence_term.args:
+                if func_cond[1].func is Ge or func_cond[1].func is Gt or func_cond[1] == True:
```
I am a little worried about this. What would happen if the original lower limit were ``-oo`` and the index ``sym`` were replaced with ``-sym``. May we assume that ``Piecewise`` would be in a standard form even after ``xreplace``?

gxyd replied:
> May we assume that Piecewise would be in a standard form even after xreplace?

As i see this now, no. I think i should add `simplify()` function on the `sequence_term` in order to make it in standard form. I will add a test case for it.

gxyd commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,181 @@ def _eval_simplify(self, ratio, measure):
+            return True
+
+        # transform sym -> -sym and swap the upper_limit = S.Infinity and lower_limit = - upper_limit
+        if lower_limit is S.NegativeInfinity:
```
Here i guess i should change this to
```
if lower_limit is S.NegativeInfinity and upper_limit is not S.Infinity:
```
?

jksuom replied:
Sums extending from ``-oo`` to ``oo`` should probably be handled by dividing them into two pieces that are considered separately.

gxyd replied:
Yes, that also sounds good to me.

gxyd commented:
@jksuom is there anything more to be done, if not then i will squash the commits. Please let me know after you review. Thanks.

jksuom commented:
Dirichlet test appears to assume a product of two arguments. Perhaps it should be tested. Maybe the test could be also run with swapped arguments.

gxyd commented:
I will look into `Dirichlet`. Do you think i should look into using `method` as an argument in `is_convergent(method='comparison')`, in which user can specify the `method` user wants to `test`, by default `method=""` hence default all tests will be run until one is conclusive. Otherwise if user specifies a particular `test`, then the result will be whether `True`, `False` or `inconclusive`.

I think you also gave some hint to this something in our gitter chat?

jksuom commented:
I don't think it would be necessary to add the keyword ``method`` now. It can be added later if desired.

gxyd commented:
I have now used it as an internal function to check for `Dirichlet test`.

Commit:
NotImplementedError for more than one symbol convergence check

test @XFAILED for Dirichlet tests

SymPy logic used instead of Python logic

test removed from dirichlet test

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+        """
+        Convergence tests are used for checking the convergence of
+        a series. There are various tests employed to check the convergence,
+        returns True if convergent and false if divergent and NotImplementedError
```
false -> False

gxyd replied:
Acknowledged

On Mon, Oct 5, 2015 at 1:23 PM, Kalevi Suominen <notifications@github.com>
wrote:

> In sympy/concrete/summations.py
> <https://github.com/sympy/sympy/pull/9906#discussion_r41117708>:
>
> > @@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
> >      def _eval_summation(self, f, x):
> >          return None
> >
> > +    def is_convergent(self):
> > +        """
> > +        Convergence tests are used for checking the convergence of
> > +        a series. There are various tests employed to check the convergence,
> > +        returns True if convergent and false if divergent and NotImplementedError
>
> false -> False
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9906/files#r41117708>.
>

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+        ### ------------- alternating series test ----------- ###
+        d = symbols('d', cls=Dummy)
+        dict_val = sequence_term.match((-1)**p*q)
+        if dict_val[p] == sym or dict_val[p].has(sym):
```
This ``has(sym)`` may allow too much, for example ``2*sym``. It would be safer to take just ``sym + const``.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+        if dict_val[p] == sym or dict_val[p].has(sym):
+            if is_decreasing(dict_val[q], interval):
+                return True
+            return False
```
The series may be convergent even if is not decreasing. I think None should be returned here.

gxyd replied:
Yes, you are right.

On Mon, Oct 5, 2015 at 1:38 PM, Kalevi Suominen <notifications@github.com>
wrote:

> In sympy/concrete/summations.py
> <https://github.com/sympy/sympy/pull/9906#discussion_r41118849>:
>
> > +        ### ----------- root test ---------------- ###
> > +        lim = Limit(abs(sequence_term)**(1/sym), sym, S.Infinity)
> > +        lim_evaluated = lim.doit()
> > +        if lim_evaluated.is_number:
> > +            if lim_evaluated < 1:
> > +                return True
> > +            if lim_evaluated > 1:
> > +                return False
> > +
> > +        ### ------------- alternating series test ----------- ###
> > +        d = symbols('d', cls=Dummy)
> > +        dict_val = sequence_term.match((-1)**p*q)
> > +        if dict_val[p] == sym or dict_val[p].has(sym):
> > +            if is_decreasing(dict_val[q], interval):
> > +                return True
> > +            return False
>
> The series may be convergent even if is not decreasing. I think None
> should be returned here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9906/files#r41118849>.
>

gxyd replied:
As a general note, `is_convergent()` would not return `None` but `NotImplementedError`. In case it can not determine. There is no case in which it would return `None`

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+        if order.expr == sequence_term:
+            pass
+        else:
+            return Sum(order.expr, (sym, lower_limit, upper_limit)).is_convergent()
```
Is is necessary to repeat the above steps with ``order.expr``? What would be different if we just replace the ``sequence_term`` with ``order.expr`` and continue?

gxyd replied:
No, that's un-necessary. I forgot to remove this code. Very thank you for
this

On Mon, Oct 5, 2015 at 1:43 PM, Kalevi Suominen <notifications@github.com>
wrote:

> In sympy/concrete/summations.py
> <https://github.com/sympy/sympy/pull/9906#discussion_r41119236>:
>
> > +        if log_n_test is not None:
> > +            if log_n_test[p] > 1:
> > +                return True
> > +            return False
> > +
> > +        # (1/(n*log(n)*log(log(n))*p)) comparison
> > +        log_log_n_test = order.expr.match(1/(sym*(log(sym)*log(log(sym))**p)))
> > +        if log_log_n_test is not None:
> > +            if log_log_n_test[p] > 1:
> > +                return True
> > +            return False
> > +
> > +        if order.expr == sequence_term:
> > +            pass
> > +        else:
> > +            return Sum(order.expr, (sym, lower_limit, upper_limit)).is_convergent()
>
> Is is necessary to repeat the above steps with order.expr? What would be
> different if we just replace the sequence_term with order.expr and
> continue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9906/files#r41119236>.
>

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+                    return False
+            except NotImplementedError:
+                pass
+            return True
```
Is this safe? Perhaps this should be moved before the ``except`` statement.

jksuom commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -855,3 +855,67 @@ def test_indexed_idx_sum():
+    assert Sum(2**n/factorial(n), (n, 1, oo)).is_convergent() == S.true
+
+    # integral test --
+    assert Sum(1/(n*log(n)), (n, 2, oo)).is_convergent() == S.false
```
This appears again below. Perhaps comparison test will catch it first.

jksuom commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -855,3 +855,67 @@ def test_indexed_idx_sum():
+
+    # alternating series tests --
+    assert Sum((-1)**(n - 1)/(n**2 - 1), (n, 3, oo)).is_convergent() == S.true
+    assert Sum((-1)**n*n, (n, 3, oo)).is_convergent() == S.false
```
This should be resolved in the first test (limit == 0). Perhaps the test should be taken with the absolute value.

gxyd replied:
I have now made use of `abs` in `divergent test`.

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+            a_n, b_n = order.expr.args[0], order.expr.args[1]
+
+            def _dirichlet_test(f_n, g_n):
+                if is_decreasing(f_n, interval):
```
This test could be taken before calling ``_dirichlet_test``.

gxyd replied:
Though since `is_decreasing` is a part of `Dirichlet test`, so i thought it
to be better to keep it inside the function itself. None the less i have
changed as you said. Let me know what you think.

On Mon, Oct 5, 2015 at 3:05 PM, Kalevi Suominen <notifications@github.com>
wrote:

> In sympy/concrete/summations.py
> <https://github.com/sympy/sympy/pull/9906#discussion_r41125572>:
>
> > +        if is_decreasing(sequence_term, interval):
> > +            integral_val = Integral(sequence_term, (sym, lower_limit, upper_limit))
> > +            try:
> > +                integral_val_evaluated = integral_val.doit()
> > +                if integral_val_evaluated.is_infinite:
> > +                    return False
> > +            except NotImplementedError:
> > +                pass
> > +            return True
> > +
> > +        ### -------------- Dirichlet tests -------------- ###
> > +        if order.expr.is_Mul:
> > +            a_n, b_n = order.expr.args[0], order.expr.args[1]
> > +
> > +            def _dirichlet_test(f_n, g_n):
> > +                if is_decreasing(f_n, interval):
>
> This test could be taken before calling _dirichlet_test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9906/files#r41125572>.
>

jksuom commented on sympy/concrete/summations.py:
```diff
@@ -253,6 +257,202 @@ def _eval_simplify(self, ratio, measure):
+            def _dirichlet_test(f_n, g_n):
+                if is_decreasing(f_n, interval):
+                    try:
+                        ing_val = integrate(g_n, (sym, interval.inf, interval.sup))
```
This is not correct in general. Though the (partial) sums of series with decreasing terms can be estimated by means of integrals, it does not follow that the same could be done with series of more general type. This line should be replaced with a more secure test for the boundedness of partial sums. (A primitive one would suffice at this stage.)

gxyd replied:
Okay, i did not knew that.So i am now making use of `Sum` instead of `integral` but since the `Sum` is not evaluated for most of these cases, leaving with no all `test case` for `Dirichlet test` in `@XFAIL`.

jksuom commented:
To make the Dirichlet test useful we should have a function for deciding whether a sequence (of partial sums) is bounded. I think that could be another (not quite easy) project. Currently, it may be best to make Dirichlet tests with ``sin`` terms XFAILed. (Not too much should be included in this PR.)

gxyd commented:
Do you feel that we should return `S.true` and `S.false` as for now, there is no 3-valued logic "True, False, None" being used, instead `NotImplementedError` is raised for not known?

asmeurer commented:
I think we should return S.true and S.false to open the way for returning convergence conditions for summations with symbolic parameters in the future (like `abs(p) < 1` for `Sum(1/x**p, (x, 1, oo))`). And I agree that NotImplementedError is better than None for unknown cases.

jksuom commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -855,3 +855,67 @@ def test_indexed_idx_sum():
+    # dirichlet tests
+    assert Sum(sin(n)/n, (n, 1, oo)).is_convergent() is S.true
+    assert Sum(sin(2*n)/n, (n, 1, oo)).is_convergent() is S.true
+    assert Sum(sin(n)/n, (n, 1, oo)).is_absolute_convergent() is S.false
```
I don't think we have any test for proving this last absolute convergence test. In particular, if Dirichlet's test fails, it does not suffice to prove that the series would be divergent. Perhaps this line should be omitted.

jksuom commented:
This looks good to me now. Some squashing would be needed. Though not all in one commit.

gxyd commented:
> Some squashing would be needed. Though not all in one commit.

I have squashed them in appropriate commits.

jksuom commented:
Ok, thanks. I'll merge in 24 hours if no one objects.

gxyd commented:
Thank you very much @jksuom . It's been really good working on this PR.

gxyd commented:
Thanks @jksuom :) .

On Thu, Oct 8, 2015 at 12:38 AM, Kalevi Suominen <notifications@github.com>
wrote:

> Merged #9906 <https://github.com/sympy/sympy/pull/9906>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9906#event-429503007>.
>

jksuom commented:
You're welcome.
